### PR TITLE
Support non-consecutive CPUs

### LINKED ISSUE #308: CPU Usage Fails when CPU Numbers aren't Consecutive
I have a Gemini running Debian which disables CPUs to save battery. When the CPUs are enabled they aren't necessarily re-enabled in order so /proc/stat might look like:

cpu ...
cpu0 ...
cpu4 ...
cpu5 ...
...

print_cpu_usage expects the cpu lines to be contain consecutively numbered cpus so it doesn't work on the Gemini. I am currently running a work around which requires you to specify the max number of CPUs in the config so isn't great (but works okay).

### PR DESCRIPTION
**Pull request by TwentyFourD**
Potential fix for Gemini PDA de-activating random CPUs

Fixes #308

### CODE REVIEW

**Comment by orestisf1993** [comment]
The checks fail because of the code formatting. You'll need to run `clang-format -i src/print_cpu_usage.c`.

Then `git add`, `git commit --amend` , `git push --force`

**Comment by TwentyFourD** [comment]
Thanks again :)

**Comment by Airblader** [comment]
It seems to me this PR is mixing a feature (new placeholder) into a bugfix. Unless I'm missing why this placeholder is crucial for the bugfix?

My main issue is that it's missing documentation, but I'd rather separate the fix from the feature in any case.

**Comment by TwentyFourD** [comment]
Yes - you're absolutely right. I'd forgotten that I'd added the feature (and not very well as it's Linux specific ATM). It's just the removal of three lines and a "}" to take out the feature. If I do that with another commit in my local branch and then do another git add, git commit --amend, git push --force, will that fix this PR in a useful way, or would it be better to create a new PR?

**Comment by Airblader** [comment]
Force pushing as you described will be totally fine :-)

**Comment by TwentyFourD** [comment]
One day I'll get the hang of git...
Thank you all for your patience. Hopefully I have it right this time :).

File: src/print_cpu_usage.c [hunk 1]
```diff
@@ -71,19 +72,30 @@ void print_cpu_usage(yajl_gen json_gen, char *buffer, const char *format, const
         curr_cpus = (struct cpu_usage *)calloc(cpu_count, sizeof(struct cpu_usage));
     }
 
+    memcpy(curr_cpus, prev_cpus, cpu_count * sizeof(struct cpu_usage));
```
Comment (orestisf1993) [inline src/print_cpu_usage.c]: Why are you copying `prev_cpus` to `curr_cpus` and then repopulating `curr_cpus`?

**Comment by TwentyFourD** [comment]
I do feel a little guilty about the multiple copying this ends up doing so I'm just testing out a slightly different way of doing the updates. This goes something like:
for (int i = 0; i < numberOfPhysicalCPUs; ++i) {
  if (cpu[i] isOnline) update(curr_sps[i]);
  else curr_cpus[i] = prev_cpus[i];
}
It isn't quite that simple, but this is the effect, and it is the same effect as I'm achieving above by the memcpy(curr_cpus, prev_cpus...) followed by updating just the online CPUs in the loop. But the code ends up looking much easier to get "off-by-one" errors in compared to the "brute-force" approach above.

File: src/print_cpu_usage.c [hunk 2]
```diff
@@ -71,19 +72,30 @@ void print_cpu_usage(yajl_gen json_gen, char *buffer, const char *format, const
...
+        if (cpu_idx < 0 || cpu_idx >= cpu_count)
             goto error;
-        curr_cpus[cpu_idx].total = curr_cpus[cpu_idx].user + curr_cpus[cpu_idx].nice + curr_cpus[cpu_idx].system + curr_cpus[cpu_idx].idle;
+        curr_cpus[cpu_idx].user = user;
+        curr_cpus[cpu_idx].nice = nice;
+        curr_cpus[cpu_idx].system = system;
+        curr_cpus[cpu_idx].idle = idle;
+        curr_cpus[cpu_idx].total = user + nice + system + idle;
+    }
+    for (int cpu_idx = 0; cpu_idx < cpu_count; cpu_idx++) {
```
[the remark this answers is not in the document]
Comment (TwentyFourD) [reply src/print_cpu_usage.c]: The disabled CPUs aren't in /proc/stat (which is why the cpu ids are non-consecutive, so they don't get updated. By the end of the first loop, the active CPUs have updated curr_cpu, but the inactive ones haven't been updated. Then curr_cpu for the inactive ones is the same as prev_cpu so they end up not contributing.

**Comment by TwentyFourD** [comment]
I've attached the code with my proposed logic so that you can see it. It is more efficient, may be more understandable, but is more complex. For code which is running 100% of the time that you are logged in, I like simple and safe even if it is a bit less efficient. However, the newer code seems to be working well, and will at least serve to illustrate a bit more how this works.
BTW, this may be a bit of an edge case. The same device, running termux under android never activates non-consecutive CPUs. However, the proposed code (should) work in all cases so is worth consideration still.
Another benefit of the proposed fix is that you can monitor cpus which are turned on and off using %cpuN. I think that will fail if you want to monitor (say) all 4 CPUs but only two are active with the current code - but I haven't verified this...
[print_cpu_usage.c.zip](https://github.com/i3/i3status/files/2425320/print_cpu_usage.c.zip)

File: src/print_cpu_usage.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (orestisf1993) [reply src/print_cpu_usage.c]: Ok, had some misunderstandings about /proc/stat

**Comment by orestisf1993** [comment]
> The same device, running termux under android never activates non-consecutive CPUs

You can manually disable them: `echo 0 | sudo tee /sys/devices/system/cpu/cpu1/online`

Haven't read the alternative code yet, will return to it. In the meantime, prefer adding a new branch to your git repo and linking that instead of using zip files.

**Comment by orestisf1993** [comment]
I like the current approach better. I doubt that the second approach is performing better than a memcpy but even if it did, the difference would be negligible since we are talking about a tiny amount of data.

> Another benefit of the proposed fix is that you can monitor cpus which are turned on and off using %cpuN. I think that will fail if you want to monitor (say) all 4 CPUs but only two are active with the current code - but I haven't verified this...

Why? Seems to be working correctly on my system.

**Pull request merged**